tools/astarte_export/lib/export.ex

```diff
@@ -0,0 +1,348 @@
+    end
+  end
+
+  defp do_process_interface(conn, realm, %{type: :individual} = interface_info, fd, state) do
+    %{
+      attributes: attributes,
+      mappings: mappings
+     } = interface_info
+
+    table_page_sizes = Application.get_env(:xandra, :cassandra_table_page_sizes, [])
+    page_size = Keyword.get(table_page_sizes, :individual_datastreams, 1000)
+    opts = [page_size: page_size]
+
+    with {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"interface", attributes}, state),
+         {:ok, state} <-
+           process_individual_streams(conn, realm, mappings, interface_info, fd, state, opts),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      {:ok, state}
+    end
+  end
+
+  defp do_process_interface(conn, realm, %{type: :object} = interface_info, fd, state) do
+    %{
+      attributes: attributes,
+      mappings: mappings
+     } = interface_info
+
+    table_page_sizes = Application.get_env(:xandra, :cassandra_table_page_sizes, [])
+    page_size = Keyword.get(table_page_sizes, :object_datastreams, 1000)
+    opts = [page_size: page_size]
+
+    with {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"interface", attributes}, state),
+         {:ok, state} <-
+           process_object_streams(conn, realm, mappings, interface_info, fd, state, opts),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      {:ok, state}
+    end
+  end
+
+
+  defp process_object_streams(conn, realm, mappings, interface_info, fd, state, opts) do
+    [h | _t] = mappings
+    fullpath = h.endpoint
+    [_, endpointprefix, _] = String.split(fullpath, "/")
+    path = "/" <> endpointprefix
+
+    sub_paths_info =
+      Enum.reduce(mappings, [], fn mapping, acc1 ->
+        path = mapping.endpoint
+        [_, _, suffix] = String.split(path, "/")
+        data_type = mapping.value_type
+        [%{suffix_path: suffix, data_type: data_type} | acc1]
+      end)
+
+    updated_interface_info =
+      Map.put(interface_info, :path, path)
+      |> Map.put(:sub_path_info, sub_paths_info)
+
+    with {:ok, state} <- XMLGenerate.xml_write_start_tag(fd, {"datastream", [path: path]}, state),
+         {:ok, state} <-
+           do_process_object_streams(conn, realm, updated_interface_info, fd, state, opts),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      {:ok, state}
+    end
+  end
+
+
+  defp process_individual_streams(_, _, [], _, _, state, _) do
+    {:ok, state}
+  end
+
+  defp process_individual_streams(conn, realm, [h | t], interface_info, fd, state, opts) do
+    with {:ok, state} <-
+           XMLGenerate.xml_write_start_tag(fd, {"datastream", [path: h.endpoint]}, state),
+         {:ok, state} <-
+           do_process_individual_streams(conn, realm, h, interface_info, fd, state, opts),
+         {:ok, state} <- XMLGenerate.xml_write_end_tag(fd, state) do
+      process_individual_streams(conn, realm, t, interface_info, fd, state, opts)
+    end
+  end
+
+
```

Comment:
Please, use `Enum.reduce_while` here too.

Reply:
I had look into this, I feel we have to continue with functional call, instead of reduce_while, as it would complicate the code structuring. 

Code structuring looks better with functional calls itself